TK Onboarding Exercise - Implement CRUD API for Recipes

clicktravel-antonrand opened the pull request:
There are a few commits for this PR - initial setup/config followed by the model and endpoint support.

clicktravel-antonrand committed:
Initial project setup

Initial project setup of a django application with docker support, linting and travis CI integration

clicktravel-antonrand committed:
Create "core" application with Recipe and Ingredients model

Create core application with a model for representing Recipes and Ingredients.

Core application contains a "wait_for_db" command to ensure PostgreSQL connectivity before starting Django application.

clicktravel-antonrand committed:
Create "recipe" application with API endpoints

Creates a recipe application containing the required CRUD operations for the exercise.

clicktravel-antonrand committed:
Update README with project details.

clicktravel-antonrand committed:
Add more context to test descriptions

clicktravel-antonrand committed:
Resolve linting issue - line too long.

RandyM1 commented on app/recipe/tests/test_recipe_api.py:
```diff
@@ -0,0 +1,255 @@
+    def test_get_all_recipes_with_name_filter_case_insensitive(self):
+        """Test retrieving a list of filtered recipes is case insensitive"""
+        # Given
+        sample_recipe()
```
Maybe add a comment here what the default name is and also below assert the name of the returned recipe.

clicktravel-antonrand replied:
I've added a comment for the default name, I'm not sure I follow:

> assert the name of the returned recipe

I'm currently asserting the recipe returned was the serialised "Turkey bubble & squeak" recipe which includes the name but can change to just check on the name instance if that's what you mean?

RandyM1 replied:
yes, so it is explicitly checking for the name value, readable in the unit test for clarity. Just my opinion how to make the test more readable, trustworthy.

RandyM1 commented on app/recipe/tests/test_recipe_api.py:
```diff
@@ -0,0 +1,255 @@
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(len(ingredients), 1)
```
Instead of `.all()` on the query set + `len()` here you can replace `all()` with `count()` - it will make a single query returning the count value instead of fetching the objects from the DB - much more efficient / clear. [You can review other query set methods here](https://docs.djangoproject.com/en/3.2/ref/models/querysets/)

clicktravel-antonrand replied:
Thank you, I'll replace Line 225 to have `ingredients.count()` but I'll keep the `.all()` on line 224 as it's needed for an assertion on line 226.

RandyM1 commented on app/recipe/views.py:
```diff
@@ -0,0 +1,19 @@
+
+class RecipeViewSet(viewsets.ModelViewSet):
+    """Manage recipes in the database"""
+    queryset = Recipe.objects.all()
```
Is this required if `get_queryset` is defined?

clicktravel-antonrand replied:
Without this I was getting the following error:

```
AssertionError: `basename` argument not specified, and could not automatically determine the name from the viewset, as it does not have a `.queryset` attribute.
```

I've just had a look in more detail and I can remove this if I add a `basename` in `recipe/urls.py` as the URL routing breaks

`router.register('', views.RecipeViewSet, basename='recipe')`

clicktravel-antonrand replied:
I've added the basename and removed the static queryset 👍

RandyM1 commented on app/core/models.py:
```diff
@@ -0,0 +1,23 @@
+
+class Recipe(models.Model):
+    """Recipe Model"""
+    name = models.CharField(max_length=255, blank=False)
```
Because we use Postgres, unless you want to limit the length, use `TextField` instead of `CharField` (it is otherwise equivalent, without requiring `max_length`

RandyM1 reviewed (approved):
Overall looks good!

clicktravel-antonrand committed:
Update model to use TextField instead of CharField

Because we use postgres these are the equivalent without the length restrictions.

clicktravel-antonrand committed:
Remove static queryset from RecipeViewSet

clicktravel-antonrand committed:
Use queryset "count" method over len

clicktravel-antonrand committed:
Comment to explain default recipe name

RandyM1 reviewed (approved):
:clap: thanks!
